packages/babel-plugin-proposal-decorators/src/transformer-2021-12.ts

```diff
@@ -0,0 +1,946 @@
+import type { NodePath } from "@babel/traverse";
+import { types as t } from "@babel/core";
+import syntaxDecorators from "@babel/plugin-syntax-decorators";
+import ReplaceSupers from "@babel/helper-replace-supers";
+import * as charCodes from "charcodes";
+
+type ClassDecoratableElement =
+  | t.ClassMethod
+  | t.ClassPrivateMethod
+  | t.ClassProperty
+  | t.ClassPrivateProperty
+  | t.ClassAccessorProperty;
+
+type ClassElement =
+  | ClassDecoratableElement
+  | t.TSDeclareMethod
+  | t.TSIndexSignature
+  | t.StaticBlock;
+
+function incrementId(id: number[], idx = id.length - 1): void {
+  // If index is -1, id needs an additional character, unshift A
+  if (idx === -1) {
+    id.unshift(charCodes.uppercaseA);
+    return;
+  }
+
+  const current = id[idx];
+
+  if (current === charCodes.uppercaseZ) {
+    // if current is Z, skip to a
+    id[idx] = charCodes.lowercaseA;
+  } else if (current === charCodes.lowercaseZ) {
+    // if current is z, reset to A and carry the 1
+    id[idx] = charCodes.uppercaseA;
+    incrementId(id, idx - 1);
+  } else {
+    // else, increment by one
+    id[idx] = current + 1;
+  }
+}
+
+/**
+ * Generates a new element name that is unique to the given class. This can be
+ * used to create extra class fields and methods for the implementation, while
+ * keeping the length of those names as small as possible. This is important for
+ * minification purposes, since public names cannot be safely renamed/minified.
+ *
+ * Names are split into two namespaces, public and private. Static and non-static
+ * names are shared in the same namespace, because this is true for private names
+ * (you cannot have #x and static #x in the same class) and it's not worth the
+ * extra complexity for public names.
+ */
+function createPrivateUidGeneratorForClass(
+  classPath: NodePath<t.ClassDeclaration | t.ClassExpression>,
+): () => t.PrivateName {
+  const currentPrivateId = [charCodes.uppercaseA];
+  const privateNames = new Set<string>();
+
+  classPath.traverse({
+    PrivateName(path) {
+      privateNames.add(path.node.id.name);
+    },
+  });
+
+  return (): t.PrivateName => {
+    let reifiedId = String.fromCharCode(...currentPrivateId);
+
+    while (privateNames.has(reifiedId)) {
+      incrementId(currentPrivateId);
+      reifiedId = String.fromCharCode(...currentPrivateId);
+    }
+
+    incrementId(currentPrivateId);
```

Comment:
Nit, but we could compact this code a bit by doing
```js
let reifiedId;
do {
  reifiedId = String.fromCharCode(...currentPrivateId);
  incrementId(currentPrivateId);
} while (privateNames.has(reifiedId));
```
and `currentPrivateId` starts from `[]` rather than `[charCodes.uppercaseA]`.